desub-current/src/metadata/mod.rs

```diff
@@ -48,13 +49,21 @@ pub enum MetadataError {
 }
 
 /// This is a representation of the SCALE encoded metadata obtained from a substrate
-/// node. While not very useful on its own, It can be passed to [`crate::Decoder`]
-/// to allow that to decode extrinsics compatible with the substrate node that
-/// this was obtained from.
+/// node. While not very useful on its own, It can be passed to [`crate::decoder`] functions
+/// to decode encoded extrinsics and storage keys.
 #[derive(Debug)]
 pub struct Metadata {
+	/// Details about the extrinsic format.
 	extrinsic: MetadataExtrinsic,
-	pallets: HashMap<u8, MetadataPallet>,
+	/// Hash pallet calls by index, since when decoding, we'll have the pallet/call
+	/// `u8`'s available to us to look them up by.
+	pallet_calls_by_index: HashMap<u8, MetadataPalletCalls>,
```

Comment:
How big is `MetadataPalletCalls`? This could be something like `[Option<MetadataPalletCalls>; 256]` (or boxed version of it) for fast `O(1)` indexing.

Could also store them in a `Vec<MetadataPalletCalls>` and then map to the indices in the vec with `[u8; 256]` or some such. Not _really_ necessary, kind of falls into premature optimization land.

`BTreeMap` would be also worth trying.

Reply:
I'd have to check, but it's not super small! This did come up in a previous PR actually too. I might have a play after lunch with your second idea; constant time lookup would be nice.

I'm not really sure why a `BTreeMap` would be more appealing than a `HashMap` here?

Reply:
I just find the idea of hashing a `u8` into `u64` and then bucketing that a bit of an overkill. Not that `BTreeMap` is any less complex necessarily, but having a key that's just an integer seems like a perfect use case for it. I have no benches to back anything, just a hunch.

Reply:
Yeah, I was thinking about the hashing after I commented above, and I can def see the appeal! Though I'm tempted to just go the full hog and do the `[u8; 256]` thing though for nice quick access :)

Reply:
@maciejhirsz there we go: https://github.com/paritytech/desub/pull/75/commits/ea34667578a8b4cf6e18128c47452c49794517da#diff-605d800671ef792606f7a7347033b612517d4d64ee4783570544eb8b3461eb51